extensions/amp-story-360/0.1/amp-story-360.js

```diff
@@ -202,15 +273,173 @@ export class AmpStory360 extends AMP.BaseElement {
     container.appendChild(this.canvas_);
     this.applyFillContent(container, /* replacedContent */ true);
 
-    return Services.storyStoreServiceForOrNull(this.win).then(
-      (storeService) => {
-        storeService.subscribe(
-          StateProperty.PAGE_SIZE,
-          this.resizeRenderer_.bind(this),
-          false /* callToInitialize */
-        );
+    attr('controls') === 'gyroscope' && this.checkGyroscopePermissions_();
+
+    Services.storyStoreServiceForOrNull(this.win).then((storeService) => {
+      this.storeService_ = storeService;
+
+      storeService.subscribe(StateProperty.PAGE_SIZE, () =>
+        this.resizeRenderer_()
+      );
+
+      storeService.subscribe(
+        StateProperty.GYROSCOPE_PERMISSION_STATE,
+        (permissionState) => this.onPermissionState_(permissionState)
+      );
+    });
+  }
+
+  /**
+   * @param {string} permissionState
+   * @private
+   */
+  onPermissionState_(permissionState) {
+    if (permissionState === 'granted') {
+      this.enableGyroscope_();
+    } else if (permissionState === 'denied') {
+      this.gyroscopeControls_ = false;
+      this.togglePermissionClass_(true);
+    }
+  }
+
+  /** @private */
+  checkGyroscopePermissions_() {
+    //  Browser doesn't support DeviceOrientationEvent API.
+    if (typeof DeviceOrientationEvent === 'undefined') {
+      return;
+    }
+
+    // If browser doesn't require permission, enable gyroscope.
+    if (typeof DeviceOrientationEvent.requestPermission === 'undefined') {
+      this.enableGyroscope_();
+    }
+
+    // If DeviceOrientationEvent and permissions, build permission button.
+    if (typeof DeviceOrientationEvent.requestPermission === 'function') {
+      this.buildPermissionButton_();
+    }
+  }
+
+  /**
+   * Creates a device orientation listener and sets gyroscopeControls_ state.
+   * If listener is not called in 1000ms, remove listener and resume animation.
+   * This happens on desktop browsers that support deviceorientation but aren't in motion.
+   * @private
+   */
+  enableGyroscope_() {
+    this.gyroscopeControls_ = true;
+    this.togglePermissionClass_(true);
+
+    this.win.addEventListener('deviceorientation', (e) => {
+      this.isReady_ && this.onDeviceOrientation_(e);
```

Comment:
Can we find a way to only do this if the component is visible on the active story page? If I understand correctly, this would render the image even when not visible as long as `layoutCallback` ran?

Reply:
Good catch. It would update every every 360 component on motion. 
We could check it `this.isReady_` and `this.isPlaying_` like we do for animate.